central/cve/telemetry/metrics_map.go

```diff
@@ -0,0 +1,111 @@
+package telemetry
+
+import (
+	"strings"
+)
+
+type aggregationKey string // e.g. "Severity|IsFixable"
+
+var opNames = map[rune]string{
+	'=': "_eq_",
+	'!': "_not_",
+	'>': "_gt_",
+	'<': "_lt_",
+}
+
+// makeMetricName sanitizes the metric key according to the Prometheus naming
+// rules, which is [a-zA-Z0-9_]+ (no colon).
+//
+// Example:
+//
+//	"CVSS>5,Cluster=*prod": "CVSS_gt_5_Cluster_eq__prod_total"
+func makeMetricName(key aggregationKey) metricName {
+	result := strings.Builder{}
+	for _, u := range key {
+		if u >= 'a' && u <= 'z' || u >= 'A' && u <= 'Z' || u >= '0' && u <= '9' {
+			result.WriteRune(u)
+		} else if u != ' ' {
+			if op, ok := opNames[u]; ok {
+				result.WriteString(op)
+			} else {
+				result.WriteRune('_')
+			}
+		}
+	}
+	result.WriteString("_total")
+	return result.String()
+}
+
+// parseAggregationExpressions parses the aggregation configuraion, and builds a
+// map, that associates every aggregation key to the list of vulnerability
+// filtering expressions.
+//
+// Example:
+//
+//	"Namespace=abc,Severity": map[metricName][]expression{
+//	  "Namespace_eq_abc_Severity_total": {"Namespace=abc", "Severity"},
```

Comment:
I am not fully sure what is being aggregated here, but it feels that prometheus provides the constructs to achieve this.

`group (metric_name{namespace="abc"}) by (severity)` or `sum (metric_name{namespace="abc"}) by (severity)`

I would appreciate an explanation why we cannot do it this way.

Reply:
We do not want to expose everything for later aggregation in prometheus because of volumes. We want, though, to aggregate something according to the user configuration. So this is a mixed approach: prometheus will aggregate on top of what is aggregated here.